File: src/pramp/AbsoluteValueSort.java
package pramp;

import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;

public class AbsoluteValueSort {
    static int[] absSort2(int[] arr) {
        // Convert int[] to Integer[] to use custom Comparator
        Integer[] arrObj = new Integer[arr.length];
        for (int i = 0; i < arr.length; i++) {
            arrObj[i] = arr[i];
        }

        // Sort using custom Comparator for absolute values
        Arrays.sort(arrObj, Comparator.comparingInt(Math::abs));

        // Convert back to int[]
        for (int i = 0; i < arr.length; i++) {
            arr[i] = arrObj[i];
        }

        return arr;
    }

    static int[] absSort(int[] arr) {
        // your code goes here
        if (arr.length <= 1) return arr;

        // 1
        Map<Integer, Integer> negMap = new HashMap<>();
        for (int v : arr) {
            if (v < 0)
                negMap.put(-v, negMap.getOrDefault(-v, 0) + 1);
        }

        System.out.println(negMap);

        // 2.make every element abs
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] < 0)
                arr[i] = Math.abs(arr[i]);
        }

        // 3.
        Arrays.sort(arr);
        System.out.println(Arrays.toString(arr));

        // 4.
        // 0 2 2 2 7
        int newVal = arr[0];
        int i = 1;

        while (i < arr.length) { //i 2->  3
            // check if the value has changed
            if (arr[i] != newVal) { // i: 0 2 -> 4
                newVal = arr[i]; // 2 -> 7

                if (negMap.containsKey(newVal)) {
                    int negCount = negMap.get(arr[i]); // count = 2 -> 1
                    while (negCount > 0) {
                        arr[i] = -arr[i];
                        negCount--; // 1 -> 0
                        i++; // 2 -> 3 -> 5
                    }
                }
            } else {
                i++;  // i 3 -> 4
            }

        }

        return arr;
    }

    public static void main(String[] args) {
        System.out.println(Arrays.toString(absSort2 (new int[]{-2, -2, 2, 5, -7})));
        System.out.println(Arrays.toString(absSort2 (new int[]{-2, -2, 2, 5, -7})));
    }
}

File: src/pramp/AbsoluteValueSort.kt
package pramp

import java.util.PriorityQueue
import kotlin.math.abs

/**
 * The custom comparator provided here, first compares the absolute
 * values of the elements. If the absolute values are equal, it then
 * compares the elements themselves to ensure that negative numbers
 * come before positive numbers when their absolute values are the same.
 */
fun absSort(arr: IntArray): IntArray {
    // Use a single priority queue with a custom comparator for absolute values
    /*val absHeap = PriorityQueue<Int> { a, b ->
        val absCompare = abs(a).compareTo(abs(b))
        if (absCompare != 0) absCompare else a.compareTo(b)
    }*/
    val absHeap = PriorityQueue<Int> { a, b ->
        if(abs(a) == abs(b)) a - b
        else abs(a) - abs(b)
    }

    // Add all elements to the heap
    for (v in arr) {
        absHeap.add(v)
    }

    // Poll from the heap to get the sorted order based on absolute values
    val result = IntArray(arr.size)
    for (i in arr.indices) {
        result[i] = absHeap.poll()
    }
    return result
}

fun absSort2(arr: IntArray): IntArray {
    val posMinHeap = PriorityQueue<Int>()
    val negMaxHeap = PriorityQueue<Int> { a, b -> b - a }

    for (v in arr) {
        if (v < 0) negMaxHeap.add(v)
        else posMinHeap.add(v)
    }

    val result = IntArray(arr.size)
    var i = 0
    while (posMinHeap.isNotEmpty() || negMaxHeap.isNotEmpty()) {
        result[i++] = when {
            posMinHeap.peek() == null -> negMaxHeap.poll()
            negMaxHeap.peek() == null -> posMinHeap.poll()
            abs(negMaxHeap.peek()) <= posMinHeap.peek() -> negMaxHeap.poll()
            else -> posMinHeap.poll()
        }
    }
    return result
}

fun absSort3(arr: IntArray): IntArray {
    arr.sort()

    var midIndex = arr.binarySearch(0)
    if (midIndex < 0) midIndex = -midIndex - 1

    val result = IntArray(arr.size)
    var p = midIndex - 1
    var q = midIndex
    var count = 0
    while (p >= 0 && q < arr.size) {
        val val1 = arr[p]
        val val2 = arr[q]
        if (val1 == -val2) {
            result[count] = val1
            p--
        } else {
            result[count] = val2
            q++
        }
        count++
    }
    return result
}

fun main() {
    println(absSort(intArrayOf(2, -7, -2, -2, 0)).contentToString())
    println(absSort(intArrayOf(-2, -1)).contentToString())
    println(absSort(intArrayOf(0, 1, 2)).contentToString())
    println(absSort(intArrayOf(2, -1, -1, -1)).contentToString())
    println(absSort(intArrayOf(-2, 3, 5, -1, 4)).contentToString())

    println(absSort(intArrayOf(-2, -1, 0, 1, 2)).joinToString(", ")) // Output: 0, -1, 1, -2, 2
}